In [2]:
import pandas as pd
book = pd.read_csv('2025도서랭킹_최종수집본.csv')

In [5]:
book

no  ranking                      bookname               authors  \
0        1        1              소년이 온다 :한강 장편소설                지은이: 한강   
1        2        2             홍학의 자리 :정해연 장편소설               지은이: 정해연   
2        3        3            작별하지 않는다 :한강 장편소설                지은이: 한강   
3        4        4                 모순 :양귀자 장편소설               지은이: 양귀자   
4        5        5  나는 소망한다 내게 금지된 것을 :양귀자 장편소설               지은이: 양귀자   
...    ...      ...                           ...                   ...   
4994  4996     4995                        까만 토끼   필리파 레더스 글·그림 ;최지현 옮김   
4995  4997     4995                   오싹오싹 좀비 금붕어     글: 모 오하라 ;번역: 지혜연   
4996  4998     4995                        봄이 오리               지은이: 한연진   
4997  4999     4995                  아빠, 나한테 물어봐    버나드 와버 글 ;이수지 그림·옮김   
4998  5000     4995           살인자의 쇼핑몰 :강지영 장편소설               지은이: 강지영   

      publisher publication_year         isbn13  addition_symbol    vol  \
0            창비             2014  9788936434120           3810.0    NaN   
1          문학동네             2021  9788954681155           3810.0    NaN   
2          문학동네             2021  9788954682152           3810.0    NaN   
3            쓰다             2013  9788998441012           3810.0    NaN   
4            쓰다             2019  9788998441074           3810.0    NaN   
...         ...              ...            ...              ...    ...   
4994    알에이치코리아           2013.0  9788925550459          73840.0    NaN   
4995        예림당           2022.0  9788930264914          74840.0    1.0   
4996  보림(보림출판사)           2025.0  9788943317799          77810.0    NaN   
4997        비룡소           2015.0  9788949113609          74800.0  234.0   
4998      자음과모음           2020.0  9788954442145           3810.0    1.0   

      class_no  ...                                       bookImageURL  \
0       813.62  ...  http://image.aladin.co.kr/product/4086/97/cove...   
1       813.70  ...  https://image.aladin.co.kr/product/27587/94/co...   
2       813.62  ...  https://image.aladin.co.kr/product/27877/5/cov...   
3       813.62  ...  https://bookthumb-phinf.pstatic.net/cover/071/...   
4       813.62  ...  https://image.aladin.co.kr/product/18936/52/co...   
...        ...  ...                                                ...   
4994    843.00  ...  https://bookthumb-phinf.pstatic.net/cover/072/...   
4995    843.00  ...  https://image.aladin.co.kr/product/29086/31/co...   
4996    813.70  ...  https://image.aladin.co.kr/product/35938/15/co...   
4997    843.00  ...  https://bookthumb-phinf.pstatic.net/cover/096/...   
4998    813.70  ...  https://image.aladin.co.kr/product/23178/43/co...   

                                     bookDtlUrl loan_count  \
0     https://data4library.kr/bookV?seq=2037399       6300   
1     https://data4library.kr/bookV?seq=6050062       5667   
2     https://data4library.kr/bookV?seq=6086446       5540   
3     https://data4library.kr/bookV?seq=1938053       5478   
4     https://data4library.kr/bookV?seq=5395822       4336   
...                                         ...        ...   
4994  https://data4library.kr/bookV?seq=1976287        331   
4995  https://data4library.kr/bookV?seq=6425211        331   
4996  https://data4library.kr/bookV?seq=7021806        331   
4997  https://data4library.kr/bookV?seq=2887808        331   
4998  https://data4library.kr/bookV?seq=5591724        331   

                                                  d_url sales_point   price  \
0     https://www.aladin.co.kr/shop/wproduct.aspx?It...     767,918  13,500   
1     https://www.aladin.co.kr/shop/wproduct.aspx?It...      53,420  12,600   
2     https://www.aladin.co.kr/shop/wproduct.aspx?It...     486,356  15,120   
3     https://www.aladin.co.kr/shop/wproduct.aspx?It...     277,357  11,700   
4     https://www.aladin.co.kr/shop/wproduct.aspx?It...      66,025  13,500   
...                                                 ...         ...     ...   
4994  https://www.a

In [4]:
book['d_url'][1]

'https://www.aladin.co.kr/shop/wproduct.aspx?ItemId=275879438'

In [6]:
book['d_url'][5]

'https://www.aladin.co.kr/shop/wproduct.aspx?ItemId=276926308'

# 알라딘 웹 크롤링
- 가격, 별점, 포인트 등등 (정적인 페이지.)
    - requests 패키지로 html을 쉽게 호출했을 때 수집가능
    - requests.get(url) 코드 한 줄로 html 가져올 수 있었다. 
    - 그리고 가져온 html에서 내가 원하는 태그를 찾았다 (find, find_all)

- 도서 상세페이지에서 **성별연령 정보를 추출하고 시도했지만 실패**
    - 단순 requests로 추출한 html에서는 찾을 수 없었다.

In [7]:
# 데이터로드
import pandas as pd     
book = pd.read_csv('2025도서랭킹_최종수집본.csv')

### **그래서, 동적인 페이지를 크롤링 하는 selenium으로 시도해보기로 한다.**
- 사전준비: 패키지, 드라이버 세팅

In [12]:
'''패키지 로드'''
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.by import By
from webdriver_manager.chrome import ChromeDriverManager
import time
from bs4 import BeautifulSoup

options = webdriver.ChromeOptions()
##
'''크롬 브라우저 동작방식 설정'''
# options.add_argument('--headless') # 활성화 하면 크롬창 안보임
options.add_argument("--start-maximized") # 브라우저 크기 최대
options.add_experimental_option("detach", True) # 프로그램이 종료되도 창 유지
options.add_argument("--disable-blink-features=AutomationControlled")
# options.add_argument("user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.102 Safari/537.36")
##

In [13]:
## 사전준비: 패키지, 드라이버 세팅
def START(): ...
START()

In [14]:
## url 주소준비
url = book['d_url'][0]
print(url)

https://www.aladin.co.kr/shop/wproduct.aspx?ItemId=40869703


In [15]:
## 드라이버 생성 및 접속체크
driver = webdriver.Chrome(options=options)
driver.get(url)

### 첫 시도 - 아무 작업 없이 html 요청
- driver에게 html 요청하는 코드는
- html = driver.page_source
- soup = BeautifulSoup(html, 'html.parser')

In [16]:
# 크롬으로 부터 html을 요청하는 코드는 아래 두 줄 고정
html = driver.page_source
soup = BeautifulSoup(html, 'html.parser')

soup.find('div', 'buyer_area_a')

<div class="buyer_area_a"><div class="buyer_box"><div class="anay_title"><div class="anay_Tround_L">구매자 분포</div><div class="anay_Tround_S">구매자<br/>분포</div></div><div class="anay_conts"><div class="analysis"><div class="analysis_box"><table border="0" cellpadding="0" cellspacing="0" width="100%"><tbody><tr><td class="tb_woman"><div class="bar" id="bar_w_0" style="width: 3.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">2.2%</div></td><td class="tb_tit">10대</td><td class="tb_man"><div class="bar" id="bar_m_0" style="width: 0.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">0.4%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_1" style="width: 23.19%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">14.6%</div></td><td class="tb_tit">20대</td><td class="tb_man"><div class="bar" id="bar_m_1" style="width: 5.59%;"><img src="//image.aladin.co.kr/

### 뭔가 html이 보이긴 하지만 아직 확실하지 않다!

In [ ]:
# 출력물을 복사 해보자
print(soup.find('div', 'buyer_area_a'))

<div class="buyer_area_a"><div class="buyer_box"><div class="anay_title"><div class="anay_Tround_L">구매자 분포</div><div class="anay_Tround_S">구매자<br/>분포</div></div><div class="anay_conts"><div class="analysis"><div class="analysis_box"><table border="0" cellpadding="0" cellspacing="0" width="100%"><tbody><tr><td class="tb_woman"><div class="bar" id="bar_w_0" style="width: 3.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">2.2%</div></td><td class="tb_tit">10대</td><td class="tb_man"><div class="bar" id="bar_m_0" style="width: 0.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">0.4%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_1" style="width: 23.19%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">14.6%</div></td><td class="tb_tit">20대</td><td class="tb_man"><div class="bar" id="bar_m_1" style="width: 5.59%;"><img src="//image.aladin.co.kr/

<div class="buyer_area_a"><div class="buyer_box"><div class="anay_title"><div class="anay_Tround_L">구매자 분포</div><div class="anay_Tround_S">구매자<br/>분포</div></div><div class="anay_conts"><div class="analysis"><div class="analysis_box"><table border="0" cellpadding="0" cellspacing="0" width="100%"><tbody><tr><td class="tb_woman"><div class="bar" id="bar_w_0" style="width: 3.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">2.2%</div></td><td class="tb_tit">10대</td><td class="tb_man"><div class="bar" id="bar_m_0" style="width: 0.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">0.4%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_1" style="width: 23.19%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">14.6%</div></td><td class="tb_tit">20대</td><td class="tb_man"><div class="bar" id="bar_m_1" style="width: 5.59%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">3.5%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_2" style="width: 26.73%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">16.8%</div></td><td class="tb_tit">30대</td><td class="tb_man"><div class="bar" id="bar_m_2" style="width: 7.64%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">4.8%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_3" style="width: 36.8%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">23.1%</div></td><td class="tb_tit">40대</td><td class="tb_man"><div class="bar" id="bar_m_3" style="width: 10.97%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">6.9%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_4" style="width: 22.85%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">14.4%</div></td><td class="tb_tit">50대</td><td class="tb_man"><div class="bar" id="bar_m_4" style="width: 11.01%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">6.9%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_5" style="width: 4.84%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">3.0%</div></td><td class="tb_tit">60대<img class="Ere_MB_m1" src="https://image.aladin.co.kr/img/shop/2021/over_sixty.png"/></td><td class="tb_man"><div class="bar" id="bar_m_5" style="width: 5.22%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">3.3%</div></td></tr><tr><td class="tb_woman_last Ere_PT10 Ere_str">여성</td><td class="tb_tit"></td><td class="tb_man_last Ere_PT10 Ere_str">남성</td></tr></tbody></table></div></div></div> </div></div>

### **과연 스크롤을 굳이 내려야 데이터가 수집될까?**

In [23]:
## 드라이버 생성 및 접속체크
driver = webdriver.Chrome(options=options)
driver.get(url)

# 크롬으로 부터 html을 요청하는 코드는 아래 두 줄 고정
html = driver.page_source
soup = BeautifulSoup(html, 'html.parser')

soup.find('div', 'buyer_area_a')

In [24]:
print(soup.find('div', 'buyer_area_a'))

None


### 굳이 스크롤을 내려야 함 
- 스크롤을 내리면 html이 변경 됨 -> 동적인 페이지
- 그럼 html 다시 호출해야 함

In [25]:
# n번만 스크롤 내리기
def scroll_n_times(driver, n, interval=0.02):
    body = driver.find_element("tag name", "body")
    for _ in range(n):
        body.send_keys("\ue00f")  # PAGE_DOWN
        time.sleep(interval)

# 사용 예시
scroll_n_times(driver, n=30)

### 스크롤을 내렸으니까 html 호출 한 번 다시하고 바로 태그도 찾아보자

In [26]:
html = driver.page_source
soup = BeautifulSoup(html, 'html.parser')

soup.find('div', 'buyer_area_a')

<div class="buyer_area_a"><div class="buyer_box"><div class="anay_title"><div class="anay_Tround_L">구매자 분포</div><div class="anay_Tround_S">구매자<br/>분포</div></div><div class="anay_conts"><div class="analysis"><div class="analysis_box"><table border="0" cellpadding="0" cellspacing="0" width="100%"><tbody><tr><td class="tb_woman"><div class="bar" id="bar_w_0" style="width: 3.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">2.2%</div></td><td class="tb_tit">10대</td><td class="tb_man"><div class="bar" id="bar_m_0" style="width: 0.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">0.4%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_1" style="width: 23.19%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">14.6%</div></td><td class="tb_tit">20대</td><td class="tb_man"><div class="bar" id="bar_m_1" style="width: 5.59%;"><img src="//image.aladin.co.kr/

<div class="buyer_area_a"><div class="buyer_box"><div class="anay_title"><div class="anay_Tround_L">구매자 분포</div><div class="anay_Tround_S">구매자<br/>분포</div></div><div class="anay_conts"><div class="analysis"><div class="analysis_box"><table border="0" cellpadding="0" cellspacing="0" width="100%"><tbody><tr><td class="tb_woman"><div class="bar" id="bar_w_0" style="width: 3.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">2.2%</div></td><td class="tb_tit">10대</td><td class="tb_man"><div class="bar" id="bar_m_0" style="width: 0.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">0.4%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_1" style="width: 23.19%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">14.6%</div></td><td class="tb_tit">20대</td><td class="tb_man"><div class="bar" id="bar_m_1" style="width: 5.59%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">3.5%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_2" style="width: 26.73%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">16.8%</div></td><td class="tb_tit">30대</td><td class="tb_man"><div class="bar" id="bar_m_2" style="width: 7.64%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">4.8%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_3" style="width: 36.8%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">23.1%</div></td><td class="tb_tit">40대</td><td class="tb_man"><div class="bar" id="bar_m_3" style="width: 10.97%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">6.9%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_4" style="width: 22.85%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">14.4%</div></td><td class="tb_tit">50대</td><td class="tb_man"><div class="bar" id="bar_m_4" style="width: 11.01%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">6.9%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_5" style="width: 4.84%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">3.0%</div></td><td class="tb_tit">60대<img class="Ere_MB_m1" src="https://image.aladin.co.kr/img/shop/2021/over_sixty.png"/></td><td class="tb_man"><div class="bar" id="bar_m_5" style="width: 5.22%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">3.3%</div></td></tr><tr><td class="tb_woman_last Ere_PT10 Ere_str">여성</td><td class="tb_tit"></td><td class="tb_man_last Ere_PT10 Ere_str">남성</td></tr></tbody></table></div></div></div> </div></div>

In [ ]:
book.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4999 entries, 0 to 4998
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   no                4999 non-null   int64  
 1   ranking           4999 non-null   int64  
 2   bookname          4999 non-null   object 
 3   authors           4999 non-null   object 
 4   publisher         4941 non-null   object 
 5   publication_year  4983 non-null   object 
 6   isbn13            4999 non-null   int64  
 7   addition_symbol   4973 non-null   float64
 8   vol               1591 non-null   float64
 9   class_no          4999 non-null   float64
 10  class_nm          4999 non-null   object 
 11  bookImageURL      4870 non-null   object 
 12  bookDtlUrl        4999 non-null   object 
 13  loan_count        4999 non-null   int64  
 14  d_url             4999 non-null   object 
 15  sales_point       4855 non-null   object 
 16  price             4855 non-null   object 


In [28]:
book

no  ranking                      bookname               authors  \
0        1        1              소년이 온다 :한강 장편소설                지은이: 한강   
1        2        2             홍학의 자리 :정해연 장편소설               지은이: 정해연   
2        3        3            작별하지 않는다 :한강 장편소설                지은이: 한강   
3        4        4                 모순 :양귀자 장편소설               지은이: 양귀자   
4        5        5  나는 소망한다 내게 금지된 것을 :양귀자 장편소설               지은이: 양귀자   
...    ...      ...                           ...                   ...   
4994  4996     4995                        까만 토끼   필리파 레더스 글·그림 ;최지현 옮김   
4995  4997     4995                   오싹오싹 좀비 금붕어     글: 모 오하라 ;번역: 지혜연   
4996  4998     4995                        봄이 오리               지은이: 한연진   
4997  4999     4995                  아빠, 나한테 물어봐    버나드 와버 글 ;이수지 그림·옮김   
4998  5000     4995           살인자의 쇼핑몰 :강지영 장편소설               지은이: 강지영   

      publisher publication_year         isbn13  addition_symbol    vol  \
0            창비             2014  9788936434120           3810.0    NaN   
1          문학동네             2021  9788954681155           3810.0    NaN   
2          문학동네             2021  9788954682152           3810.0    NaN   
3            쓰다             2013  9788998441012           3810.0    NaN   
4            쓰다             2019  9788998441074           3810.0    NaN   
...         ...              ...            ...              ...    ...   
4994    알에이치코리아           2013.0  9788925550459          73840.0    NaN   
4995        예림당           2022.0  9788930264914          74840.0    1.0   
4996  보림(보림출판사)           2025.0  9788943317799          77810.0    NaN   
4997        비룡소           2015.0  9788949113609          74800.0  234.0   
4998      자음과모음           2020.0  9788954442145           3810.0    1.0   

      class_no  ...                                       bookImageURL  \
0       813.62  ...  http://image.aladin.co.kr/product/4086/97/cove...   
1       813.70  ...  https://image.aladin.co.kr/product/27587/94/co...   
2       813.62  ...  https://image.aladin.co.kr/product/27877/5/cov...   
3       813.62  ...  https://bookthumb-phinf.pstatic.net/cover/071/...   
4       813.62  ...  https://image.aladin.co.kr/product/18936/52/co...   
...        ...  ...                                                ...   
4994    843.00  ...  https://bookthumb-phinf.pstatic.net/cover/072/...   
4995    843.00  ...  https://image.aladin.co.kr/product/29086/31/co...   
4996    813.70  ...  https://image.aladin.co.kr/product/35938/15/co...   
4997    843.00  ...  https://bookthumb-phinf.pstatic.net/cover/096/...   
4998    813.70  ...  https://image.aladin.co.kr/product/23178/43/co...   

                                     bookDtlUrl loan_count  \
0     https://data4library.kr/bookV?seq=2037399       6300   
1     https://data4library.kr/bookV?seq=6050062       5667   
2     https://data4library.kr/bookV?seq=6086446       5540   
3     https://data4library.kr/bookV?seq=1938053       5478   
4     https://data4library.kr/bookV?seq=5395822       4336   
...                                         ...        ...   
4994  https://data4library.kr/bookV?seq=1976287        331   
4995  https://data4library.kr/bookV?seq=6425211        331   
4996  https://data4library.kr/bookV?seq=7021806        331   
4997  https://data4library.kr/bookV?seq=2887808        331   
4998  https://data4library.kr/bookV?seq=5591724        331   

                                                  d_url sales_point   price  \
0     https://www.aladin.co.kr/shop/wproduct.aspx?It...     767,918  13,500   
1     https://www.aladin.co.kr/shop/wproduct.aspx?It...      53,420  12,600   
2     https://www.aladin.co.kr/shop/wproduct.aspx?It...     486,356  15,120   
3     https://www.aladin.co.kr/shop/wproduct.aspx?It...     277,357  11,700   
4     https://www.aladin.co.kr/shop/wproduct.aspx?It...      66,025  13,500   
...                                                 ...         ...     ...   
4994  https://www.a

### chatbot에게 명령
### 내가 테스트하고 성공한 코드

driver = webdriver.Chrome(options=options)
driver.get(url)

scroll_n_times(driver, n=30)

html = driver.page_source
soup = BeautifulSoup(html, 'html.parser')

soup.find('div', 'buyer_area_a')

### soup.find 결과물
<div class="buyer_area_a"><div class="buyer_box"><div class="anay_title"><div class="anay_Tround_L">구매자 분포</div><div class="anay_Tround_S">구매자<br/>분포</div></div><div class="anay_conts"><div class="analysis"><div class="analysis_box"><table border="0" cellpadding="0" cellspacing="0" width="100%"><tbody><tr><td class="tb_woman"><div class="bar" id="bar_w_0" style="width: 3.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">2.2%</div></td><td class="tb_tit">10대</td><td class="tb_man"><div class="bar" id="bar_m_0" style="width: 0.57%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">0.4%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_1" style="width: 23.19%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">14.6%</div></td><td class="tb_tit">20대</td><td class="tb_man"><div class="bar" id="bar_m_1" style="width: 5.59%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">3.5%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_2" style="width: 26.73%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">16.8%</div></td><td class="tb_tit">30대</td><td class="tb_man"><div class="bar" id="bar_m_2" style="width: 7.64%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">4.8%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_3" style="width: 36.8%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">23.1%</div></td><td class="tb_tit">40대</td><td class="tb_man"><div class="bar" id="bar_m_3" style="width: 10.97%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">6.9%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_4" style="width: 22.85%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">14.4%</div></td><td class="tb_tit">50대</td><td class="tb_man"><div class="bar" id="bar_m_4" style="width: 11.01%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">6.9%</div></td></tr><tr><td class="tb_woman"><div class="bar" id="bar_w_5" style="width: 4.84%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">3.0%</div></td><td class="tb_tit">60대<img class="Ere_MB_m1" src="https://image.aladin.co.kr/img/shop/2021/over_sixty.png"/></td><td class="tb_man"><div class="bar" id="bar_m_5" style="width: 5.22%;"><img src="//image.aladin.co.kr/img/shop/2018/analysis_bg1.png"/></div> <div class="per">3.3%</div></td></tr><tr><td class="tb_woman_last Ere_PT10 Ere_str">여성</td><td class="tb_tit"></td><td class="tb_man_last Ere_PT10 Ere_str">남성</td></tr></tbody></table></div></div></div> </div></div>

# 명령
- 데이터프레임 book에 d_url이 실제 각 페이지의 주소이다.
- 남성에서 가장 높은 %를 보이는 연령대, 여성에서 가장 높은 %를 보이는 연령대를 추출해서 
'남성_top_연령대', '여성_top_연령대'에 추가.
- 첫 3개 페이지만 테스트 해보고 싶다.
- 패키지 로드와 드라이버 세팅 코드 필요없음.

### 데이터프레임에 직접 적용해보기 위한 코드를 챗봇으로부터 추출(첫 3개 행만)

In [34]:
driver = webdriver.Chrome(options=options)
driver.get(url)

# 결과 저장용 컬럼 초기화
book['남성_top_연령대'] = None
book['여성_top_연령대'] = None

# 앞 3개 페이지만 테스트
for idx in book.index[:3]:
    url = book.loc[idx, 'd_url']
    driver.get(url)

    scroll_n_times(driver, n=30)

    soup = BeautifulSoup(driver.page_source, 'html.parser')

    buyer_area = soup.find('div', class_='buyer_area_a')
    if buyer_area is None:
        continue

    rows = buyer_area.find_all('tr')

    male_data = {}
    female_data = {}

    for row in rows:
        age_td = row.find('td', class_='tb_tit')
        woman_td = row.find('td', class_='tb_woman')
        man_td = row.find('td', class_='tb_man')

        if not (age_td and woman_td and man_td):
            continue

        age = age_td.get_text(strip=True)

        woman_per = woman_td.find('div', class_='per')
        man_per = man_td.find('div', class_='per')

        if woman_per and man_per:
            female_data[age] = float(woman_per.get_text(strip=True).replace('%', ''))
            male_data[age] = float(man_per.get_text(strip=True).replace('%', ''))

    # 최댓값 연령대 추출
    if male_data:
        book.loc[idx, '남성_top_연령대'] = max(male_data, key=male_data.get)

    if female_data:
        book.loc[idx, '여성_top_연령대'] = max(female_data, key=female_data.get)

In [33]:
book

no  ranking                      bookname               authors  \
0        1        1              소년이 온다 :한강 장편소설                지은이: 한강   
1        2        2             홍학의 자리 :정해연 장편소설               지은이: 정해연   
2        3        3            작별하지 않는다 :한강 장편소설                지은이: 한강   
3        4        4                 모순 :양귀자 장편소설               지은이: 양귀자   
4        5        5  나는 소망한다 내게 금지된 것을 :양귀자 장편소설               지은이: 양귀자   
...    ...      ...                           ...                   ...   
4994  4996     4995                        까만 토끼   필리파 레더스 글·그림 ;최지현 옮김   
4995  4997     4995                   오싹오싹 좀비 금붕어     글: 모 오하라 ;번역: 지혜연   
4996  4998     4995                        봄이 오리               지은이: 한연진   
4997  4999     4995                  아빠, 나한테 물어봐    버나드 와버 글 ;이수지 그림·옮김   
4998  5000     4995           살인자의 쇼핑몰 :강지영 장편소설               지은이: 강지영   

      publisher publication_year         isbn13  addition_symbol    vol  \
0            창비             2014  9788936434120           3810.0    NaN   
1          문학동네             2021  9788954681155           3810.0    NaN   
2          문학동네             2021  9788954682152           3810.0    NaN   
3            쓰다             2013  9788998441012           3810.0    NaN   
4            쓰다             2019  9788998441074           3810.0    NaN   
...         ...              ...            ...              ...    ...   
4994    알에이치코리아           2013.0  9788925550459          73840.0    NaN   
4995        예림당           2022.0  9788930264914          74840.0    1.0   
4996  보림(보림출판사)           2025.0  9788943317799          77810.0    NaN   
4997        비룡소           2015.0  9788949113609          74800.0  234.0   
4998      자음과모음           2020.0  9788954442145           3810.0    1.0   

      class_no  ... loan_count  \
0       813.62  ...       6300   
1       813.70  ...       5667   
2       813.62  ...       5540   
3       813.62  ...       5478   
4       813.62  ...       4336   
...        ...  ...        ...   
4994    843.00  ...        331   
4995    843.00  ...        331   
4996    813.70  ...        331   
4997    843.00  ...        331   
4998    813.70  ...        331   

                                                  d_url sales_point   price  \
0     https://www.aladin.co.kr/shop/wproduct.aspx?It...     767,918  13,500   
1     https://www.aladin.co.kr/shop/wproduct.aspx?It...      53,420  12,600   
2     https://www.aladin.co.kr/shop/wproduct.aspx?It...     486,356  15,120   
3     https://www.aladin.co.kr/shop/wproduct.aspx?It...     277,357  11,700   
4     https://www.aladin.co.kr/shop/wproduct.aspx?It...      66,025  13,500   
...                                                 ...         ...     ...   
4994  https://www.aladin.co.kr/shop/wproduct.aspx?It...         402   9,000   
4995  https://www.aladin.co.kr/shop/wproduct.aspx?It...         175  10,800   
4996  https://www.aladin.co.kr/shop/wproduct.aspx?It...       1,084  13,500   
4997  https://www.aladin.co.kr/shop/wproduct.aspx?It...       3,761  12,600   
4998  https://www.aladin.co.kr/shop/wproduct.aspx?It...       2,137  12,420   

     page_count   category review_count  average_rating 남성_top_연령대  여성_top_연령대  
0         216.0       한국소설        394.0             9.6        40대         40대  
1         336.0  추리/미스터리소설         53.0             7.3        20대         30대  
2         332.0       한국소설        110.0             9.0        50대         40대  
3         308.0       한국소설        107.0             9.1       None        None  
4         368.0       한국소설         38.0             9.4       None        None  
...         ...        ...          ...             ...        ...         ...  
4994       40.0        그림책          8.0             9.3       None        None  
4995      192.0   동화/명작/고전         14.0            10.0       None        None  
4996       32.0        그림책          9.0            10.0       None        None  
4997       40.0        그림책         23.

In [35]:
book.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4999 entries, 0 to 4998
Data columns (total 23 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   no                4999 non-null   int64  
 1   ranking           4999 non-null   int64  
 2   bookname          4999 non-null   object 
 3   authors           4999 non-null   object 
 4   publisher         4941 non-null   object 
 5   publication_year  4983 non-null   object 
 6   isbn13            4999 non-null   int64  
 7   addition_symbol   4973 non-null   float64
 8   vol               1591 non-null   float64
 9   class_no          4999 non-null   float64
 10  class_nm          4999 non-null   object 
 11  bookImageURL      4870 non-null   object 
 12  bookDtlUrl        4999 non-null   object 
 13  loan_count        4999 non-null   int64  
 14  d_url             4999 non-null   object 
 15  sales_point       4855 non-null   object 
 16  price             4855 non-null   object 
